src/rshim_pcie_lf.c

```diff
@@ -93,6 +109,201 @@ static int pci_cap_write(struct pci_dev *pci_dev, int offset, uint32_t value)
   return 0;
 }
 
+/* Acquire and release the MSN GW_CR_64B lock */
+static int msn_gw_lock_acquire(struct pci_dev *pci_dev)
+{
+  uint32_t read_value;
+  time_t t0, t1;
+  int rc;
+
+  /* Wait until the MSN GW_CR_64B lock is free */
+  time(&t0);
+  do {
+    rc = pci_cap_read(pci_dev, MSN_GW_CR_64B_NIC_FW_REG0, &read_value);
+    if (rc)
+      return rc;
+
+    time(&t1);
+    if (difftime(t1, t0) > RSHIM_LOCK_RETRY_TIME)
+      return -ETIMEDOUT;
+  } while (read_value & MSN_GW_CR_64B_NIC_FW_LOCK);
+
+  return 0;
+}
+
+static int msn_gw_lock_release(struct pci_dev *pci_dev)
+{
+  int rc;
+
+  /* Release MSN GW_CR_64B lock */
+  rc = pci_cap_write(pci_dev, MSN_GW_CR_64B_NIC_FW_REG0, 0x0);
+
+  return rc;
+}
+
+/* Poll the MSN GW_CR_64B */
+static int msn_gw_poll_busy(struct pci_dev *pci_dev)
+{
+  uint32_t read_value;
+  time_t t0, t1;
+  int rc;
+
+  /* Wait until the MSN GW_CR_64B lock is free */
+  time(&t0);
+  do {
+    rc = pci_cap_read(pci_dev, MSN_GW_CR_64B_NIC_FW_REG0_COPY, &read_value);
+    if (rc)
+      return rc;
+
+    time(&t1);
+    if (difftime(t1, t0) > RSHIM_LOCK_RETRY_TIME)
+      return -ETIMEDOUT;
+  } while (read_value & MSN_GW_CR_64B_NIC_FW_BUSY);
+
+  return 0;
+}
+
+/*
+ * Mechanism to access RShim via the MSN GW_CR_64B gateway.
+ */
+static int msn_gw_read(struct pci_dev *pci_dev, int addr,
+                               uint64_t *result)
+{
+  uint64_t read_result;
+  uint32_t data;
+  int rc;
+
+  /* Acquire MSN_GW_NIC_FW_LOCK */
+  rc = msn_gw_lock_acquire(pci_dev);
+  if (rc)
+    return rc;
+
+  /* Write addr to MSN_GW_ADDR */
+  rc = pci_cap_write(pci_dev, MSN_GW_CR_64B_NIC_FW_ADDR, (addr >> 2));
+  if (rc)
+    return rc;
```

Comment:
We should release the lock for all the cases, such as using goto here and release the lock at the end.